host/settings/pin/pin_test.go

```diff
@@ -0,0 +1,269 @@
+	mu       sync.Mutex
+	value    float64
+	currency string
+}
+
+func (e *exchangeRateRetrieverStub) updateRate(value float64) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+	e.value = value
+}
+
+func (e *exchangeRateRetrieverStub) SiacoinExchangeRate(_ context.Context, currency string) (float64, error) {
+	e.mu.Lock()
+	defer e.mu.Unlock()
+
+	if !strings.EqualFold(currency, e.currency) {
+		return 0, errors.New("currency not found")
+	}
+	return e.value, nil
+}
+
+func convertToCurrency(target decimal.Decimal, rate decimal.Decimal) types.Currency {
+	hastings := target.Div(rate).Mul(decimal.New(1, 24)).Round(0).String()
+	c, err := types.ParseCurrency(hastings)
+	if err != nil {
+		panic(err)
+	}
+	return c
+}
+
+func checkSettings(settings settings.Settings, pinned pin.PinnedSettings, expectedRate float64) error {
+	rate := decimal.NewFromFloat(expectedRate)
+	if pinned.Storage > 0 {
+		storagePrice := convertToCurrency(decimal.NewFromFloat(pinned.Storage), rate).Div64(4320).Div64(1e12)
+		if !storagePrice.Equals(settings.StoragePrice) {
+			return fmt.Errorf("expected storage price %d, got %d", storagePrice, settings.StoragePrice)
+		}
+	}
+
+	if pinned.Ingress > 0 {
+		ingressPrice := convertToCurrency(decimal.NewFromFloat(pinned.Ingress), rate).Div64(1e12)
+		if !ingressPrice.Equals(settings.IngressPrice) {
+			return fmt.Errorf("expected ingress price %d, got %d", ingressPrice, settings.IngressPrice)
+		}
+	}
+
+	if pinned.Egress > 0 {
+		egressPrice := convertToCurrency(decimal.NewFromFloat(pinned.Egress), rate).Div64(1e12)
+		if !egressPrice.Equals(settings.EgressPrice) {
+			return fmt.Errorf("expected egress price %d, got %d", egressPrice, settings.EgressPrice)
+		}
+	}
+
+	if pinned.MaxCollateral > 0 {
+		maxCollateral := convertToCurrency(decimal.NewFromFloat(pinned.MaxCollateral), rate)
+		if !maxCollateral.Equals(settings.MaxCollateral) {
+			return fmt.Errorf("expected max collateral %d, got %d", maxCollateral, settings.MaxCollateral)
+		}
+	}
+	return nil
+}
+
+func TestPinnedFields(t *testing.T) {
+	log := zaptest.NewLogger(t)
+	db, err := sqlite.OpenDatabase(filepath.Join(t.TempDir(), "test.db"), log)
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer db.Close()
+
+	rr := &exchangeRateRetrieverStub{
+		value:    1,
+		currency: "usd",
+	}
+
+	node, err := test.NewNode(t.TempDir())
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer node.Close()
+
+	sm, err := settings.NewConfigManager(settings.WithHostKey(types.GeneratePrivateKey()), settings.WithStore(db), settings.WithChainManager(node.ChainManager()))
+	if err != nil {
+		t.Fatal(err)
+	}
+	defer sm.Close()
+
+	pm, err := pin.NewManager(pin.WithAverageRateWindow(time.Minute),
+		pin.WithFrequency(100*time.Millisecond),
+		pin.WithExchangeRateRetriever(rr),
+		pin.WithSettings(sm),
+		pin.WithStore(db),
+		pin.WithLogger(log.Named("pin")))
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	initialSettings := sm.Settings()
+	pin := pin.PinnedSettings{
+		Currency: "usd",
+
+		Threshold:     0.1,
```

Comment:
`10%` is quite a lot. I'd probably do `2-3%` to also trigger in more stable market conditions. Otherwise you might miss out of a good amount of money as a host.

Reply:
The user can set it to whatever they want. I don't think it really matters in the test